app/models/project_topic.rb

```diff
@@ -0,0 +1,95 @@
+class ProjectTopic < ApplicationRecord
+  has_many :signed_up_teams, foreign_key: :sign_up_topic_id, dependent: :destroy
+  has_many :teams, through: :signed_up_teams # list all teams choose this topic, no matter in waitlist or not
+  has_many :assignment_questionnaires, class_name: 'AssignmentQuestionnaire', foreign_key: 'topic_id', dependent: :destroy
+  belongs_to :assignment
+
+  # max_choosers should be a non-negative integer
+  validates :max_choosers, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+
+  # Checks if there are any slots currently available
+  # Returns true if the number of available slots is greater than 0, otherwise false
+  def slot_available?
+    current_available_slots > 0
+  end
+
+  # Assigns the current topic to a team by updating the provided sign-up record.
+  # Marks the sign-up as not waitlisted and associates it with the current topic.
+  def assign_topic_to_team(new_sign_up)
+    new_sign_up.update(is_waitlisted: false, sign_up_topic_id: self.id)
+  end
+
+  # Adds a new sign-up to the waitlist by updating the sign-up record.
+  # Marks the sign-up as waitlisted and associates it with the current topic.
+  def save_waitlist_entry(new_sign_up)
+    new_sign_up.update(is_waitlisted: true, sign_up_topic_id: self.id)
+  end
+
+  # Signs up a team for the current topic.
+  # Checks if the team is already signed up, and if so, ensures they are not waitlisted.
+  # If a slot is available, assigns the topic to the team; otherwise, adds the team to the waitlist.
+  def sign_up_team(team_id)
+    topic_id = self.id
```

Comment:
It would be better to remove this assignment stmt and just use "self.id" in the code below.